rllib/algorithms/algorithm_config.py

```diff
@@ -2775,60 +2775,123 @@ def get_default_learner_class(self) -> Union[Type["Learner"], str]:
+                # otherwise we see if they have provided a multi-agent spec that
+                # specifies the SingleAgentRLModuleSpec to use instead of the default,
+                # in that case, we use that spec for the values. otherwise we use
+                # the default spec for the values.
+                if isinstance(
+                    cur_marl_module_spec.module_specs, SingleAgentRLModuleSpec
+                ):
+                    # The individual module specs are defined by the user
+                    single_agent_spec = module_spec or cur_marl_module_spec.module_specs
+                    module_specs = {
+                        k: copy.deepcopy(single_agent_spec) for k in policy_dict.keys()
+                    }
+                else:
+                    # The individual module specs are not defined by the user,
+                    # so we use the default
+                    single_agent_spec = module_spec or default_rl_module
+                    module_specs = {
+                        k: copy.deepcopy(
+                            cur_marl_module_spec.module_specs.get(k, single_agent_spec)
+                        )
+                        for k in policy_dict.keys()
+                    }
+
+                marl_module_spec = cur_marl_module_spec.__class__(
+                    marl_module_class=cur_marl_module_spec.marl_module_class,
+                    module_specs=module_specs,
+                )
+            else:
+                # Default is multi-agent and user wants to override it. In this case,
+                # we have two options: 1) the user provided a multi-agent spec, in
+                # which case we use that for the values, 2) self.rl_module_spec is a
+                # spec that defines SingleAgentRLModuleSpecs to be used for everything.
+                # In this case, we need to use that spec for the values.
+                if module_spec is None:
+                    if isinstance(
+                        cur_marl_module_spec.module_specs, SingleAgentRLModuleSpec
+                    ):
+                        # The individual module specs are not given, it is given as one
+                        # SingleAgentRLModuleSpec to be re-used for all
+                        single_agent_spec = cur_marl_module_spec.module_specs
+                    else:
+                        raise ValueError(
+                            "MultiAgentRLModuleSpec is given but no module_spec is "
+                            "provided when adding a policy."
+                        )
+                else:
+                    single_agent_spec = module_spec
+
+                marl_module_spec = cur_marl_module_spec.__class__(
+                    marl_module_class=cur_marl_module_spec.marl_module_class,
+                    module_specs={
+                        k: copy.deepcopy(single_agent_spec) for k in policy_dict.keys()
+                    },
```

Comment:
In the case `isinstance(default_rl_module, MultiAgentRLModuleSpec)` -> `module_spec is None` -> `isinstance( cur_marl_module_spec.module_specs, SingleAgentRLModuleSpec)`), this will make it so that we overwrite all the SARL module specs from the user-specified MARL Module spec the with the default SingleAgentRLModuleSpec.
We should use https://github.com/ray-project/ray/pull/33128/files#diff-c2f47542f9a0a91de7e62d127e13303ccfa8cf5353af44494f51a418d7d9d922R463-R483 ?

Reply:
.update was not merged back then, but it was manually happening at the end of this method. I have added a TODO to use .update() instead.

Reply:
https://github.com/kouroshHakha/ray/blob/self-play-learner/rllib/algorithms/algorithm_config.py#L2893

Reply:
I overwrite here, but then at the end go through all specs and update them with default.